test/model/internal_link_test.dart

```diff
@@ -0,0 +1,394 @@
+  }) {
+    assert((streams != null || users != null) ^ (store != null));
+    store ??= setupStore(realmUrl: realmUrl, streams: streams, users: users);
+    for (final testCase in testCases) {
+      final String urlString = testCase.$1;
+      final Uri url = realmUrl.resolve(urlString);
+      final Narrow? expected = testCase.$2;
+      test(urlString, () {
+        check(parseInternalLink(url, store!)).equals(expected);
+      });
+    }
+  }
+
+  group('ParseInternalLink', () {
+    final streams = [
+      eg.stream(streamId: 1, name: 'check'),
+    ];
+    final testCases = [
+      (true, 'legacy: stream name, no ID', realmUrl,
+        realmUrl.resolve('#narrow/stream/check')),
+      (true, 'legacy: stream name, no ID, topic', realmUrl,
+        realmUrl.resolve('#narrow/stream/check/topic/topic1')),
+
+      (true, 'with numeric stream ID', realmUrl,
+        realmUrl.resolve('#narrow/stream/123-check')),
+      (true, 'with numeric stream ID and topic', realmUrl,
+        realmUrl.resolve('#narrow/stream/123-check/topic/topic1')),
+
+      (true, 'with numeric pm user IDs (new operator)', realmUrl,
+        realmUrl.resolve('#narrow/dm/123-mark')),
+      (true, 'with numeric pm user IDs (old operator)', realmUrl,
+        realmUrl.resolve('#narrow/pm-with/123-mark')),
+
+      (false, 'wrong fragment', realmUrl,
+        realmUrl.resolve('#nope')),
+      (false, 'wrong path', realmUrl,
+        realmUrl.resolve('user_uploads/#narrow/stream/check')),
+      (false, 'wrong domain', realmUrl,
+        Uri.parse('https://another.com/#narrow/stream/check')),
+
+      (false, '#narrowly', realmUrl,
+        realmUrl.resolve('#narrowly/stream/check')),
+
+      (false, 'double slash', realmUrl,
+        Uri.parse('https://example.com//#narrow/stream/check')),
+      (false, 'triple slash', realmUrl,
+        Uri.parse('https://example.com///#narrow/stream/check')),
+
+      (true, 'with port', Uri.parse('https://example.com:444/'),
+        Uri.parse('https://example.com:444/#narrow/stream/check')),
+
+      // Dart's [Uri] currently lacks IDNA or Punycode support:
+      //   https://github.com/dart-lang/sdk/issues/26284
+      //   https://github.com/dart-lang/sdk/issues/29420
+
+      // (true, 'same domain, punycoded host', Uri.parse('https://example.भारत/'),
+      //   Uri.parse('https://example.xn--h2brj9c/#narrow/stream/check')), // FAILS
+
+      (true, 'punycodable host', Uri.parse('https://example.भारत/'),
+        Uri.parse('https://example.भारत/#narrow/stream/check')),
+
+      // (true, 'same domain, IDNA-mappable', Uri.parse('https://example.com'),
+      //   Uri.parse('https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check')), // FAILS
+
+      (true, 'ipv4 address', Uri.parse('http://192.168.0.1/'),
+        Uri.parse('http://192.168.0.1/#narrow/stream/check')),
+
+      // (true, 'same IPv4 address, IDNA-mappable', Uri.parse('http://192.168.0.1/')),
+      //   Uri.parse('http://１𝟗𝟚。①⁶🯸．₀｡𝟭/#narrow/stream/check')), // FAILS
```

Comment:
I think these are all more readable the way they were (which is how we have them in the zulip-mobile tests): the action is all happening in the authorities (the hosts + ports), particularly in comparing the realm URL's authority to the input URL's authority. So having those two aligned with each other is best for that.